cmd/bbolt/command_surgery_meta.go

```diff
@@ -0,0 +1,317 @@
+	if _, err := checkSourceDBPath(srcDBPath); err != nil {
+		return err
+	}
+
+	if err := common.CopyFile(srcDBPath, cfg.outputDBFilePath); err != nil {
+		return fmt.Errorf("[meta update] copy file failed: %w", err)
+	}
+
+	pageSize := cfg.pageSize
+	if pageSize == 0 {
+		pageSize = uint32(common.DefaultPageSize)
+	}
+
+	// update meta page 0
+	m0, buf0, err := ReadMetaPageAt(cfg.outputDBFilePath, 0, pageSize)
+	if err != nil {
+		return fmt.Errorf("read the first meta page failed: %w", err)
+	}
+	m0Changed := updateMetaField(m0, cfg.fields)
+	if m0Changed {
+		if err := writeMetaPageAt(cfg.outputDBFilePath, buf0, 0, pageSize); err != nil {
+			return fmt.Errorf("[meta update] write the first meta page failed: %w", err)
+		}
+	}
+
+	if pageSize != m0.PageSize() {
+		fmt.Fprintf(os.Stdout, "WARNING: The provided page size (%d) isn't consistent with the page size (%d) in the first valid meta page\n", pageSize, m0.PageSize())
+	}
+
+	// update meta page 1
+	m1, buf1, err := ReadMetaPageAt(cfg.outputDBFilePath, 1, pageSize)
+	if err != nil {
+		return fmt.Errorf("read the second meta page failed: %w", err)
+	}
+
+	m1Changed := updateMetaField(m1, cfg.fields)
+	if m1Changed {
+		if err := writeMetaPageAt(cfg.outputDBFilePath, buf1, 1, pageSize); err != nil {
+			return fmt.Errorf("[meta update] write the second meta page failed: %w", err)
+		}
+	}
+
+	// Display results
+	if !m0Changed && !m1Changed {
+		fmt.Fprintln(os.Stdout, "Nothing changed!")
+	}
+
+	if m0Changed {
+		fmt.Fprintln(os.Stdout, "The first meta page has been updated!")
+	}
+	if m1Changed {
+		fmt.Fprintln(os.Stdout, "The second meta page has been updated!")
+	}
+
+	return nil
+}
+
+func updateMetaField(m *common.Meta, fields []string) bool {
+	changed := false
+	for _, field := range fields {
+		kv := strings.Split(field, ":")
+		val, _ := strconv.ParseUint(kv[1], 10, 64)
+
+		switch kv[0] {
+		case "pageSize":
+			m.SetPageSize(uint32(val))
+		case "root":
+			m.SetRootBucket(common.NewInBucket(common.Pgid(val), 0))
+		case "freelist":
+			m.SetFreelist(common.Pgid(val))
+		case "pgid":
+			m.SetPgid(common.Pgid(val))
+		case "txid":
+			m.SetTxid(common.Txid(val))
+		}
+
+		changed = true
+	}
+
+	if m.Magic() != common.Magic {
+		m.SetMagic(common.Magic)
+		changed = true
+	}
+	if m.Version() != common.Version {
+		m.SetVersion(common.Version)
+		changed = true
+	}
+	if m.Flags() != common.MetaPageFlag {
+		m.SetFlags(common.MetaPageFlag)
+		changed = true
+	}
+
+	newChecksum := m.Sum64()
+	if m.Checksum() != newChecksum {
+		m.SetChecksum(newChecksum)
+		changed = true
+	}
+
+	return changed
+}
+
+func ReadMetaPageAt(dbPath string, metaPageId uint32, pageSize uint32) (*common.Meta, []byte, error) {
+	if metaPageId > 1 {
+		return nil, nil, fmt.Errorf("invalid metaPageId: %d", metaPageId)
+	}
+
+	f, err := os.OpenFile(dbPath, os.O_RDONLY, 0444)
+	if err != nil {
+		return nil, nil, err
+	}
+	defer f.Close()
+
+	buf := make([]byte, 1024)
```

Comment:
should it be page-size buf?

Reply:
It should be fine, because the meta data is just about 52 bytes.

Reply:
It is fine, but it's confusing to have arbitrary high value set. Can you just use pagesize, or leave a comment why 1024?

Reply:
Added a comment,

```
	// The meta page is just 64 bytes, and definitely less than 1024 bytes,
	// so it's fine to only read 1204 bytes. Note we don't care about the
	// pageSize when reading the first meta page, because we always read the
	// file starting from offset 0. Actually the passed pageSize is 0 when
	// reading the first meta page in the `surgery meta update` command.
```